src/lib.rs

```diff
@@ -921,34 +1014,103 @@ impl<T: TensorType> Tensor<T> {
     }
 }
 
-impl<T: TensorType> Drop for Tensor<T> {
-    fn drop(&mut self) {
-        self.drop_tensor();
+impl<T: TensorType> Deref for TensorDataNoCRepr<T> {
+    type Target = [T];
+
+    #[inline]
+    fn deref(&self) -> &[T] {
+        self.unpack();
+        unsafe { slice::from_raw_parts(self.data.get(), self.data_count) }
+    }
+}
+
+impl<T: TensorType> DerefMut for TensorDataNoCRepr<T> {
+    #[inline]
+    fn deref_mut(&mut self) -> &mut [T] {
+        self.unpack();
+        if !T::is_repr_c() {
+            // If the slice is modified, the tensor is stale.
+            self.drop_tensor();
+        }
+        unsafe { slice::from_raw_parts_mut(self.data.get(), self.data_count) }
+    }
+}
+
+/// Holds a multi-dimensional array of elements of a single data type.
+///
+/// The data buffer stores elements in row major order.  E.g. if data is treated
+/// as a vector of `T`:
+///
+/// ```text
+///   element 0:   index (0, ..., 0)
+///   element 1:   index (0, ..., 1)
+///   ...
+/// ```
+#[derive(Debug)]
+pub struct Tensor<T: TensorType> {
+    inner: <T as TensorType>::InnerType,
+    dims: Vec<u64>,
+}
+
+#[inline]
+fn product(values: &[u64]) -> u64 {
+    values.iter().product()
+}
+
+impl<T: TensorType> Tensor<T> {
+    /// Creates a new tensor.
+    ///
+    /// The data is initialized to zeros.
+    pub fn new(dims: &[u64]) -> Self {
+        Tensor {
+            inner: <T as TensorType>::InnerType::new_inner(dims),
+            dims: Vec::from(dims),
+        }
+    }
+
+    /// Sets (copies) the tensor values to the provided ones.
+    ///
+    /// ```
+    /// # use tensorflow::Tensor;
+    /// let a = Tensor::new(&[2, 2]).with_values(&[0_i32, 1, 2, 3]).unwrap();
+    /// ```
+    pub fn with_values(mut self, value: &[T]) -> Result<Self> {
+        if self.len() != value.len() {
+            return Err(invalid_arg!(
+                "length of values array ({}) is not equal to tensor total elements ({})",
+                value.len(), self.len()));
+        }
+        for (e, v) in self.iter_mut().zip(value) {
+            e.clone_from(v);
+        }
+        Ok(self)
+    }
+
+    /// Returns the tensor's dimensions.
+    pub fn dims(&self) -> &[u64] {
+        &self.dims
+    }
+
+    // Wraps a TF_Tensor. Returns None if types don't match.
+    unsafe fn from_tf_tensor(tensor: *mut tf::TF_Tensor) -> Option<Self> {
+        if DataType::from_c(tf::TF_TensorType(tensor)) != T::data_type() {
```

Comment:
Since this check is done inside `T::InnerType::from_tf_tensor`, it doesn't need to be kept here.